### ISSUE TITLE
unused_optional_binding misses ifs inside functions

### ISSUE THREAD

**Issue by keith** [realm/SwiftLint#1237]
This example does not trigger the unused_optional_binding:

```swift
func foo() {
    if let _ = bar {
    }
}
```

If you add this test case:

```swift
"func foo() {\nif let _ = bar {\n}\n"
```

You can see that the tests fail to pass. I believe this is because the `if` in this case is nested in the substructure of the func, which isn't handled by the [current logic](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/UnusedOptionalBindingRule.swift#L50)

// @rakaramos

**Comment by marcelofabri** [comment]
I think this is a problem on the default implementation of `ASTRule`: https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Protocols/ASTRule.swift#L21. It only looks at substructures of the same `KindType`.

**Comment by keith** [comment]
Ah yep! Looks believable. Do you know of any immediate problems we'd have if we "fixed" that?

**Comment by marcelofabri** [comment]
I can only think about performance regressions, but we should care more about correctness in this case 😅
